libr/anal/types.c

```diff
@@ -369,24 +369,42 @@ R_API char *r_anal_type_func_args_name(RAnal *anal, const char *func_name, int i
 	return NULL;
 }
 
-/* WTF always return 1? */
-static int captureSubString (void *p, const char *k, const char *v) {
-	char **ret = (char **)p;
-	if (strcmp (v, "func")) {
-		return 1;
+#define MIN_MATCH_LEN 4
+
+R_API char *r_anal_type_func_guess(RAnal *anal, char *func_name) {
+	int i = 0, j =0, offset = 0, n;
+	char *str = func_name;
+	char* res = NULL;
+
+	size_t slen = strlen(str);
+	if (slen < MIN_MATCH_LEN) {
+		return NULL;
 	}
-	if (strstr (ret[1], k) && (!ret[0] || strlen(k) > strlen (ret[0]))) {
-		free (ret[0]);
-		ret[0] = strdup (k);
+	if((slen > 4 && func_name[4] == '.') &&
+		((str[0] == 's' && str[1] == 'y'  && str[0] == 'm') ||
+		 (str[0] == 'f' && str[1] == 'u'  && str[0] == 'n')    )) {
+		offset = 4;
+	}
+	str = strdup(&func_name[offset]);
+	slen -= offset;
+	for (n = slen; n >= MIN_MATCH_LEN; --n) {
+		for (j = 0; j < slen - n ; ++j) {
+			char saved = str[j+n];
+			str[j+n] = 0;
+			if (sdb_exists(anal->sdb_types, &str[j])) {
+				res = sdb_get(anal->sdb_types, str, 0);
+				// XXX if we throw away all of these, why keep them
+				// in same sdb? most types are func. Wasted work.
+				bool is_func = res && !strcmp("func", res);
+				if (!is_func) {
+					goto stage_left;
```

Comment:
I actually prefer everything exit through single chokepoint, more simple to audit. but ok.